src/Components/ContextMenu.re

```diff
@@ -0,0 +1,257 @@
+};
+
+// TYPES
+
+module Id: {
+  type t;
+  let create: unit => t;
+} = {
+  type t = int;
+
+  let lastId = ref(0);
+  let create = () => {
+    incr(lastId);
+    lastId^;
+  };
+};
+
+[@deriving show({with_path: false})]
+type item('data) = {
+  label: string,
+  // icon: option(IconTheme.IconDefinition.t),
+  data: [@opaque] 'data,
+};
+
+type t('data) = {
+  id: Id.t,
+  x: int,
+  y: int,
+  originX: [ | `Left | `Middle | `Right],
+  originY: [ | `Top | `Middle | `Bottom],
+  items: list(item('data)),
+};
+
+// MENUITEM
+
+module MenuItem = {
+  module Constants = {
+    let fontSize = 12;
+  };
+
+  module Styles = {
+    open Style;
+
+    let bg = (~theme: Theme.t, ~isFocused) =>
+      isFocused ? theme.menuSelectionBackground : theme.menuBackground;
+
+    let container = (~theme, ~isFocused) => [
+      padding(10),
+      flexDirection(`Row),
+      backgroundColor(bg(~theme, ~isFocused)),
+    ];
+
+    // let icon = fgColor => [
+    //   fontFamily("seti.ttf"),
+    //   fontSize(Constants.fontSize),
+    //   marginRight(10),
+    //   color(fgColor),
+    // ];
+
+    let label = (~font: UiFont.t, ~theme: Theme.t, ~isFocused) => [
+      fontFamily(font.fontFile),
+      textOverflow(`Ellipsis),
+      fontSize(Constants.fontSize),
+      color(theme.menuForeground),
+      backgroundColor(bg(~theme, ~isFocused)),
+    ];
+  };
+
+  let make:
+    'data.
+    (
+      ~item: item('data),
+      ~theme: Theme.t,
+      ~font: UiFont.t,
+      ~onClick: unit => unit,
+      unit
+    ) =>
+    React.element(React.node)
+   =
+    (~item, ~theme, ~font, ~onClick, ()) => {
+      let isFocused = false;
+
+      // let iconView =
+      //   switch (item.icon) {
+      //   | Some(icon) =>
+      //     IconTheme.IconDefinition.(
+      //       <Text
+      //         style={Styles.icon(icon.fontColor)}
+      //         text={FontIcon.codeToIcon(icon.fontCharacter)}
+      //       />
+      //     )
+
+      //   | None => <Text style={Styles.icon(Colors.transparentWhite)} text="" />
+      //   };
+
+      let labelView = {
+        let style = Styles.label(~font, ~theme, ~isFocused);
+        <Text style text={item.label} />;
+      };
+
+      <Clickable onClick>
+        <View style={Styles.container(~theme, ~isFocused)}>
+          // iconView
+           labelView </View>
+      </Clickable>;
+      // onMouseOut={_ => setIsFocused(_ => false)}
+      // onMouseOver={_ => setIsFocused(_ => true)}
+    };
+};
+
+// MENU
+
+module Menu = {
+  module Styles = {
+    open Style;
+
+    let container = (~x, ~y, ~theme: Theme.t) => [
+      position(`Absolute),
+      top(y),
+      left(x),
+      backgroundColor(theme.menuBackground),
+      color(theme.menuForeground),
+      width(Constants.menuWidth),
+    ];
+  };
+
+  let component = React.Expert.component("Menu");
+  let make = (~model, ~theme, ~font, ~onItemSelect, ()) =>
+    component(hooks => {
+      let ((maybeRef, setRef), hooks) = Hooks.state(None, hooks);
```

Comment:
There's some unfortunate weak polymorphism issues when using the brisk-reconciler ppx, so we have to go old-school here. The ppx also has issues with type annotations, so `MenuItem` above can't use it either because it needs a type variable to be universally quantified.

Reply:
Ah... interesting. I wonder if we'll hit more issues like this. Nice catch.

Reply:
Probably. It's not the first time I've hit the type annotation issue. It's tracked here: https://github.com/briskml/brisk-reconciler/issues/53. There's also an issue with locally abstract types. Both of those are just omissions though, and should be relatively easy to fix. I'm not entirely sure about the weak polymorhpism issue as I haven't figured out the root cause and a repro yet.